File: competitive-programming/src/edu/algos/li03_hashing/easy/L1_TwoSum.java
package edu.algos.li03_hashing.easy;

import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

public class L1_TwoSum {

	public static void main(String[] args) {
		L1_TwoSum obj = new L1_TwoSum();
		System.out.println(Arrays.toString(obj.twoSum(new int[] {2,7,11,15}, 9)));
	}
	
    // Problem: https://leetcode.com/problems/two-sum/
    // Idea: Use Map to put number and its index
	public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>();
        for(int i=0; i<nums.length; i++){
            int rem = target-nums[i];
            if(map.containsKey(rem)){
                return new int[]{map.get(rem), i};
            }else{
                map.put(nums[i], i);                 
            }
        }
        return new int[]{};
    }

}

File: dsa/patterns/02_hashing/python/l1_two_sum.py
# L1 - Two Sum
#
# Problem: https://leetcode.com/problems/two-sum/
#
# Idea:
# - Use a hashmap to store number → index
# - For each number, check if (target - number) exists
#
# Time: O(n)
# Space: O(n)


def two_sum(nums, target):
    index_map = {}

    for i, num in enumerate(nums):
        complement = target - num

        if complement in index_map:
            return [index_map[complement], i]

        index_map[num] = i

    return []


# -------------------- TEST --------------------
if __name__ == "__main__":
    nums = [2, 7, 11, 15]
    target = 9
    print(two_sum(nums, target))  # Output: [0, 1]
